ldmud: various build fixes

cpu opened the pull request:
## libxcrypt-legacy

The LDMud game engine defaults to using a legacy DES hash that is (sensibly) only enabled in the `libxcrypt-legacy` derivation, not `libxcrypt`.

This commit switches LDMud to that build input instead of `libxcrypt`. This fixes a runtime error when calling `efun::crypt()` from LPC that would print a confusing error message like "crypt() is not available.".

See also https://github.com/NixOS/nixpkgs/pull/223034

## libxml2 2.12+ compat

The v2.12.0 release of `libxml2` changed the definition of the callback used with `xmlSetStructuredErrorFunc()` (ref `libxml2` [NEWS](https://gitlab.gnome.org/GNOME/libxml2/-/blob/master/NEWS)).

A fix is underway upstream (https://github.com/ldmud/ldmud/pull/95), but the release cadence is such that we should expect to fix this locally for some time, so a local patch with a fix is applied.

## mariadb-connector-c compat

The upstream `configure.ac`'s `AC_LANG_PROGRAM` for mysql feature detection seems to be incompatible with the `mariadb-connector-c` package, resulting in support not being detected. 

Digging in to the `config.log` we can see it's a type mismatch error from the test program `.c`:

```
configure:11661: checking for mySQL
configure:11694: gcc -c -g -O2 -fwrapv -I/usr/inet6/include -I/nix/store/ya8wpj6dqz39024v6xrv504i9kyidpil-mariadb-connector-c-3.1.21-dev/include/mysql conftest.c >&5
conftest.c: In function 'foo':
conftest.c:90:12: error: returning 'MYSQL *' {aka 'struct st_mysql *'} from a function with incompatible return type 'struct MYSQL *' [-Wincompatible-pointer-types]
90 |     return &var;
|            ^~~~
```

We resolves the issue locally by applying a small patch.

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

cpu committed:
ldmud: use libxcrypt-legacy

The LDMud game driver defaults to using a legacy DES hash that is
(sensibly) only enabled in the libxcrypt-legacy derivation, not
libxcrypt.

This commit switches LDMud to that build input instead of libxcrypt.
This fixes a runtime error when calling `efun::crypt()` from LPC that
would print a confusing error message like "crypt() is not available.".

cpu commented:
@vcunat Would you be willing to review this PR as the author of #223034? Thanks! :bow:

vcunat commented:
Oops.  At this point, the package won't build even with this PR.  Probably some change in libxml2 triggered that.

cpu commented:
I'll take a look

cpu committed:
ldmud: apply patch for libxml2 2.12+ compat

A fix is underway upstream, but the release cadence is such that we
should expect to fix this locally for some time.

cpu committed:
ldmud: patch mysql configure script detection

The upstream `configure.ac`'s `AC_LANG_PROGRAM` for mysql feature
detection seems to be incompatible with the `mariadb-connector-c`
package, resulting in support not being detected.

Digging in to the `config.log` we can see it's a type mismatch error
from the test program `.c`:

```
configure:11661: checking for mySQL
configure:11694: gcc -c -g -O2 -fwrapv -I/usr/inet6/include -I/nix/store/ya8wpj6dqz39024v6xrv504i9kyidpil-mariadb-connector-c-3.1.21-dev/include/mysql conftest.c >&5
conftest.c: In function 'foo':
conftest.c:90:12: error: returning 'MYSQL *' {aka 'struct st_mysql *'} from a function with incompatible return type 'struct MYSQL *' [-Wincompatible-pointer-types]
90 |     return &var;
|            ^~~~
```

This commit resolves the issue locally by applying a small patch.

cpu commented:
> At this point, the package won't build even with this PR. Probably some change in libxml2 triggered that.

@vcunat Indeed you are correct. Thanks for flagging.

I pushed two commits (one fixing the `libxml2` issue, and one fixing `ldmud-full` for `mariadb-connector-c`) and updated the PR description.

vcunat commented:
Fixes the build and looks OK at a glance, so why not.
